relayer/src/source.ts

```diff
@@ -1,56 +1,116 @@
 import { ApiPromise } from "@polkadot/api";
-import { Header, Hash, SignedBlock } from "@polkadot/types/interfaces";
 import { Observable } from "@polkadot/types/types";
-import { Text, U64 } from "@polkadot/types/primitive";
-import { concatMap } from "rxjs/operators";
+import { U64 } from "@polkadot/types/primitive";
+import { concatMap, take, map } from "rxjs/operators";
+import { from, merge } from 'rxjs';
+import { Logger } from "pino";
+import { Header, Hash, SignedBlock, Block } from "@polkadot/types/interfaces";
 
-import { TxData } from "./types";
+import { ParaHeadAndId, TxData } from "./types";
+import { getParaHeadAndIdFromRecord, isRelevantRecord } from './utils'
+import Parachain from "./parachain";
 
-type SourceParams = {
+type SourceConstructorParams = {
   api: ApiPromise;
-  chain: Text;
+  chain: string;
   feedId: U64;
+  parachainsMap: Map<string, Parachain>;
+  logger: Logger;
 };
 
 class Source {
   private api: ApiPromise;
-  private chain: Text;
+  private chain: string;
   private feedId: U64;
+  private parachainsMap: Map<string, Parachain>;
+  private logger: Logger;
 
-  constructor({ api, chain, feedId }: SourceParams) {
-    this.api = api;
-    this.chain = chain;
-    this.feedId = feedId;
+  constructor(params: SourceConstructorParams) {
+    this.api = params.api;
+    this.chain = params.chain;
+    this.feedId = params.feedId;
+    this.parachainsMap = params.parachainsMap;
+    this.logger = params.logger;
+    this.getBlocksByHeader = this.getBlocksByHeader.bind(this);
+    this.getParablocks = this.getParablocks.bind(this);
   }
 
-  private subscribeHeads = (): Observable<Header> =>
-    this.api.rx.rpc.chain.subscribeFinalizedHeads();
+  private subscribeHeads(): Observable<Header> {
+    return this.api.rx.rpc.chain.subscribeFinalizedHeads();
+  }
+
+  private getBlock(hash: Hash): Observable<SignedBlock> {
+    return this.api.rx.rpc.chain.getBlock(hash).pipe(take(1))
```

Comment:
`take` automatically unsubscribes after the first execution, otherwise, we will have to assign observable to a variable and call the `unsubscribe` method explicitly